flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java

```diff
@@ -52,8 +54,10 @@ public int getPriority() {
 		return priority;
 	}
 
-	public Runnable getRunnable() {
-		return runnable;
+	public void tryCancel(boolean mayInterruptIfRunning) {
+		if (runnable instanceof Future) {
+			((Future<?>) runnable).cancel(mayInterruptIfRunning);
+		}
```

Comment:
I'm not sure if that's the right place for this check. `MailboxExecutor` is actually creating `FutureTask` instances, which is also a kind of private implementation detail that the Runnable can be a `Future`. Maybe we should move `tryCancel` method to the `MailboxExecutor` somehow? 

Something is not right here, but this also was not right before and I'm not entirely sure how to fix it. However I'm not entirely sure if the fix that you proposed is much better. Previously `Mail` was exposing its `Runnable`, which wasn't super pretty, but also wasn't a very big deal, but with this commit and without it, this `instanceof Future` check is leaking through the code :(

Reply:
Agree, but since `runnable` is encapsulated in `Mail` then it's the responsibility of `Mail` to cancel it if possible.
One approach that comes to mind is to construct `Mail` with `closeCallback` along with `Runnable`, but that seems overcomplicated.